packages/scope-manager/src/referencer/TypeVisitor.ts

```diff
@@ -197,7 +197,21 @@ class TypeVisitor extends Visitor {
   protected TSMappedType(node: TSESTree.TSMappedType): void {
     // mapped types key can only be referenced within their return value
     this.#referencer.scopeManager.nestMappedTypeScope(node);
-    this.visitChildren(node);
+    this.#referencer
+      .currentScope()
+      .defineIdentifier(
+        node.key,
+        new TypeDefinition(
+          node.key,
+          node.parent as
+            | TSESTree.TSInterfaceDeclaration
+            | TSESTree.TSTypeAliasDeclaration
+            | TSESTree.TSTypeParameter,
+        ),
+      );
+    this.visit(node.constraint);
+    this.visit(node.nameType);
+    this.visit(node.typeAnnotation);
```

Comment:
@bradzacher I'm low confidence I understood this right... going off the other type visitors with identifier/key-like nodes, this seemed like the right way to go?...

Reply:
First up - we need to visit things in the order that makes sense from a variable reference POV.

The previous code just did `visitChildren`. This meant that the visitor would work off of the ordering defined in the `visitor-keys`, eg: `TSMappedType: ['nameType', 'typeParameter', 'typeAnnotation'],` [1] and `typeParameter` defines a variable for the `.name` and then manually visits the `.constraint` then the `.default`.

[1] Note: I don't remember why I put `nameType` first in #2748... It seems wrong 🤔 Not sure...

---

You've done a good job of re-interpreting the existing code.
But to explain what we do here in the scope analyser.

1) Visit the mapped type
2) We know it has one index signature and nothing but one index signature...
3) That signature defines a type variable (the "key") and that variable can be referenced later in the mapped type - so we need to define a type variable can be referenced by the "constraint" and the "name type".
4) then we visit the children and do whatever needs to be done for them.
    - Because we had to pull apart the `key` to create a variable - we can't just `visitChildren` automatically and have to instead manually visit each child.

So your logic is all sound.

In terms of the ordering of these calls - you've defined it in source order and that should be all good. There's no way for the `constraint` or `nameType` to declare a "variable" that can be referenced later - which means that the ordering doesn't explicitly matter - so source ordering is the best choice.

Reply:
Ok cool thanks, this makes sense I think!